Feature/doc restructure rosidl typesupport microxrcedds (#14)
microROS/micro-ROS-doc: rosidl_typesupport_microxrcedds/Code_generation.md

﻿# Code generation for C type-support

> Type support generation starting point are .msg files.

Reviewer: ```suggestion
Type support generation starting point is a set of .msg files.
```

These kind of files are the input for `rosidl_generator_c` ROS2 package which processes them and starts the type support generation during application build.

`rosidl_generator_c` take those message definitions and generates code from each one.
It generates the `<type_name>.h` that you should include to use your type.
The code files generated by `rosidl_generator_c` provides you with the .msg definition in C along with the type support for it.

This type support, in this case, is tied with the code generated by `rosidl_typessuport_c`.
`rosidl_typessuport_c` is another ROS2 package which generates functions to get the underlying middleware type support for each type.
In our case, as we support `rosidl_typesupport_microxrcedds_c`, `rosidl_typessuport_c` generates a function which makes the association between the type and the type support used for Micro XRCE-DDS.

`rosidl_typesupport_microxrcedds_c` ROS2 package generates middleware-specific functions per each .msg found.

> This middleware-specific functions in our case include serialization and desserialization using Micro-CDR as in any other Micro XRCE-DDS application.

Reviewer: ```suggestion
This middleware-specific functions in our case include serialisation and deserialisation using Micro-CDR as in any other Micro XRCE-DDS application.
```

> Those functions get stored in a `rosidl_message_type_support_t` C structure which is used for serialize and desserialize ROS2 messages later on.

Reviewer: ```suggestion
Those functions get stored in a `rosidl_message_type_support_t` C structure which is used in the ROS2 messages serialisation and deserialisation later on.
```

> As result of this process you will end up having a type representation and a type support for it. The former should be used to represent data of that type and the former are the structures which provides information to the publishers and subscribers on how to handle (serialize/desserialize) the type.

Reviewer: ```suggestion
As a result of this process, you will end up having a type representation and type support for it. The former should be used to represent data of that type, and the former are the structures which provide information to the publishers and subscribers on how to handle (serialise/deserialise) the type.
```

## Usage

To use any type, you need to include the generated `<type_name>.h` which provides you with enough utilities to create messages of that type.

> Also, you need a way to get the type support used by this type, for that purpose you have to get the type support functions. For simplicity, rclc provides a macro to help you get the right function names which gives you the required `rosidl_message_type_support_t` tied to your type.

Reviewer: ```suggestion
Also,  it would help if you had a way to get the type support used by this type, for that purpose you have to get the type support functions. For simplicity, rclc provides a macro to help you get the right function names which give you the required `rosidl_message_type_support_t` tied to your type.
```
